### PR TITLE
Optionally build with openssl instead of boringssl

### PR DESCRIPTION
**Pull request by bacek**
Currently grpc-sys builds with boringssl. Which causes issues when I try to use openssl-based crates )e.g. reqwest) in the same project. Main reason is ```rustc``` way linking of static libraries.

Building of this code fails with a lot of unresolved symbols

```rust
extern crate grpcio;
extern crate reqwest;

use std::sync::Arc;

fn main() {
    let _c = reqwest::Client::new();

    let e = Arc::new(grpcio::Environment::new(2));
    let cred = grpcio::ChannelCredentialsBuilder::new().build();
    let _channel = grpcio::ChannelBuilder::new(e).secure_connect("example.com:443", cred);

    println!("Hello, world!");
}

```

```bash
         /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:192: undefined reference to `BIO_meth_new'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:195: undefined reference to `BIO_meth_set_write'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:196: undefined reference to `BIO_meth_set_read'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:197: undefined reference to `BIO_meth_set_puts'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:198: undefined reference to `BIO_meth_set_ctrl'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:199: undefined reference to `BIO_meth_set_create'
          /home/bacek/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-0.9.24/src/ssl/bio.rs:200: undefined reference to `BIO_meth_set_destroy'

```

Building with openssl solves it issue:

Cargo.toml
```toml
[package]
name = "req-test"
version = "0.1.0"
authors = ["Vasily Chekalkin <bacek@bacek.com>"]

[dependencies.grpcio]
version = "0.2.3"
path = "../grpc-rs"
features = ["secure", "openssl"]

[dependencies]
reqwest = "0.8.5"

```

Build results:
```bash
$ cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling req-test v0.1.0 (file:///home/bacek/work/req-test)                 
    Finished dev [unoptimized + debuginfo] target(s) in 3.36 secs
```

### CODE REVIEW

**Comment by Hoverbear** [comment]
Hi @bacek !

Cool! This seems reasonable.

I'd like to make sure we properly test with this feature. Can you please add to the `.travis.yml` in the `script` tag as the last step:

```
  - cargo test --features openssl --all
```

That should let the CI properly test with this feature.

**Comment by bacek** [comment]
Build is epicly failing. @stepancheg broke/unbroke with 1.7/2.0 releases protobuf.

File: Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ members = ["proto", "benchmark", "compiler", "interop"]
 default = ["protobuf-codec"]
 protobuf-codec = ["protobuf"]
 secure = ["grpcio-sys/secure"]
+openssl = ["grpcio-sys/openssl"]
```
Comment (BusyJay) [inline Cargo.toml]: How about add `secure` features as a dependency?

Comment (bacek) [reply Cargo.toml]: Done. Also re-based vs master to fix build.

**Comment by BusyJay** [comment]
CI fails.

**Comment by bacek** [comment]
It's probably because Trusty is really old...

**Comment by siddontang** [comment]
em, do we need to install OpenSSL manually for the Rust OpenSSL?

**Comment by bacek** [comment]
https://github.com/sfackler/rust-openssl/blob/master/README.md

> rust-openssl depends on OpenSSL version 1.0.1 or above, or LibreSSL. Both the libraries and headers need to be present in the build environment before this crate is compiled, and some instructions of how to do this are in the sections below.

**Comment by kostko** [comment]
I've tried this out and it seems there are problems with OpenSSL 1.1.0 as I am getting things like the following during runtime:
```
E0611 09:42:25.911327254   19890 ssl_transport_security.c:830] Sending protected frame to ssl failed with -1
E0611 09:42:25.911341601   19890 secure_endpoint.c:185]      Decryption error: TSI_INTERNAL_ERROR
```

It seems to be related to a check in `grpc` (`ssl_protector_unprotect`) where `protected_frames_bytes_size` can be `0` and this seems to fail with an error on recent OpenSSL. Fixing this issue (e.g., see [this commit](https://github.com/ekiden/grpc/commit/dc7396fdfe89999a631b6c6fdc4b148a74dc7615)) makes the TLS connections work.

**Comment by ian-p-cooke** [comment]
I was having some rust-openssl and grpcio problems myself because I needed 1.1.0 and I solved that by building/installing grpc with EMBED_OPENSSL=false.  Based on the CI failure I think you need to set that variable when building as well because it shouldn't even be trying to compile thirdparty/boringssl if you want to use openssl.

File: grpc-sys/Cargo.toml [hunk 2]
```diff
@@ -36,6 +36,7 @@ exclude = [
 
 [dependencies]
 libc = "0.2"
+openssl = { version = "0.10", optional = true }
```
Comment (ian-p-cooke) [inline grpc-sys/Cargo.toml]: do you need this dependency at all?  you just need openssl installed, right?

File: grpc-sys/build.rs [hunk 3]
```diff
@@ -72,6 +72,9 @@ fn build_grpc(cc: &mut Build, library: &str) {
         if cfg!(target_os = "macos") {
             config.cxxflag("-stdlib=libc++");
         }
+        if cfg!(feature = "openssl") {
+            config.define("gRPC_SSL_PROVIDER", "package");
```
Comment (ian-p-cooke) [inline grpc-sys/build.rs]: as mentioned in my PR comment I think you want to define EMBED_OPENSSL=false instead.

**Pull request closed without merging**